### DISCUSSION TITLE
开发建议：「要素定制」与「范例文件」

### DISCUSSION THREAD

**Discussion by yanhuacuo** [pdog18/Soak#9] [General]
首先，一直在关注这个项目，这个项目的技术路线非常理想，但是实现效果不尽如意。表现在两个方面：

- 需要使用者足够了解 RIME

这里，一方面是各文件的功能与命名规范，另一方面是「权级」问题。其实现在各种方案实现非常混乱。

有些方案小群体几乎仅靠 「wubi98_ci.custom.yaml」等价的「方案补丁」写单行补丁来实现一切功能，还有则是互相参照，将不同的功能归在不同的文件中来实现，但是外行人看上去也是一头雾水。

如果不是对上述文件非常熟悉，那几乎无法使用「润笔」。因为有些代码，他复制走了也不知道放在哪里。

- 本地化不够

当然，看到了有本地化，app 化的开发目标，但是如果仅是将网页打包，是不够的。RIME 的方案制定太混乱了，虽然参数配置很自由，但是如果遵循一定的规范，其实并不妨碍这种自由。如果我们提前把这种规范确定下来，符合规范写法的配置皆可使用工具，就能最大化地让所有人受益。

### 范本文件内置

范本文件是某种方案能在 RIME 上实现的成熟配置，仅以「98五笔」为例：

- 文件结构

default.yaml→「挂单文件」
default.custom.yaml.yaml→「程序修改时会被重新覆写的挂单文件」
weasel.yaml→「主题文件」
weasel.custom.yaml→「程序修改时会被重新覆写的主题文件」
wubi98_ci.dict.yaml→「主词库文件
wubi98_ci.extended.dict.yaml→「扩展词库文件
wubi98_ci.custom.yaml→「方案补丁」
wubi98_ci.schema.yaml→「方案配置主文件」
hot_keys.yaml→「中英状态管理」

- schema结构

多翻译器实现引导造词；
引入Lua
引入扩展词库

### 参数修改规范

这里是最麻烦的地方，因为几乎所有的参数都可以通过写在 「wubi98_ci.custom.yaml」中，以补丁命令来实现。这样写出来的文件不仅难读，而且容易出故障，特别是单行补丁指定行数 next 的那种写法。

98五笔小组是这样处理这个问题的：

一：schema 的参数，直接在 schema 上改修；
二：方案补丁文件「wubi98_ci.custom.yaml」接管「weasel.custom.yaml」与「weasel.yaml」。
三：中英文状态管理，则通过「weasel.custom.yaml」引入「hot_keys.custom.yaml」来实现。

首先，schema 上的参数极多，功能极多，任何形式的补丁，都没有在这里直接用工具修改来得便利；
其次，「weasel.custom.yaml」与「weasel.yaml」这部分，属于不同平台下的「前端文件」，它们一般有统一的发布格式。用「wubi98_ci.custom.yaml」来接管前端参数，在功能分流上比较简明，写出来的补丁格式也更统一，方便管理：[wubi98_ci.custom.yaml](https://github.com/yanhuacuo/weasel-data/blob/main/Rime/wubi98_ci.custom.yaml)
最后，之所以在 [weasel.custom.yaml](https://github.com/yanhuacuo/weasel-data/blob/main/Rime/weasel.custom.yaml) 中引入[中英文状态文件](https://github.com/yanhuacuo/weasel-data/blob/main/Rime/hot_keys.yaml) ，是因为 weasel 和 fcitx5-rime 都是在等效文件中管理中英文状态参数的，fcitx5 则是在 fcitx5.custom.yaml 中，这样做，多平台比较统一。

## 实现

如果未来做成 APP，那么可以由用户勾选方案，「吐出范本文件」或「窗口引导功能设定」的方式来完成「方案初始化」，这样用户也不迷路了，APP 也有的放矢了，可以说是最佳解决方案。

比如，勾选「双拼」，窗口引导完成「主题设置」，「输入功能设置」，「快捷键一览」……最终完成配置。

98的示范配置：https://github.com/yanhuacuo/weasel-data

供大佬参考。

**Comment by pdog18** [comment]
非常非常非常感谢您关注此 repo，提到的意见对我也很有帮助。
---
关于您提到的

> 如果不是对上述文件非常熟悉，那几乎无法使用「润笔」。因为有些代码，他复制走了也不知道放在哪里。

这一点可能需要我简单解释一下，可能我们预设的用户状态不同导致的问题，目前及以后 soak 主要是考虑用户刚下载 Rime 还未进行任何配置的时候，直接将 soak 产出的文件放入指定文件夹。所以不是复制代码而应该是直接移动产出的文件。

当然这也是网页版的限制，无法直接将文件投放到目标文件夹（尤其是安装目录是处于‘系统’文件夹），如果是 Electron App 那么可能就没有这个问题了

---

关于您提到的桌面 app 与浏览器 app 的区别，我没能太理解您的意思，在我看来目前他们的区别只在于是否可以直接修改 Rime 的文件已经主动执行 Rime 重新部署之类的浏览器无法操作的行为。

我尝试理解您的内容，感觉您表示的是通过我们预设 feature 然后给出已实现的方案是好的行为，我同样也这样认为的！也同样明白如果只是将 yaml 的完全自由的配置转为 GUI 是完全没有意义上的（虽然 soak 按键设置就是这样糟糕）

---

另外您提到的「窗口引导功能设定」也是我考虑过的（也参考其他输入法刚安装时的行为），后续确实也会根据您的这个想法尝试进行设计

---

其他：
目前主要是几个地方阻碍了我对 soak 的开发。
1. 确实时间不太多
2. 是对目前 soak 的整体设计（表现在最终实现效果上）极度不满意，考虑完全重新调整一下
3. Rime 目前有很多不够优秀的地方，让我丧失了继续进行的兴趣（英文单词、多行文本的支持等）

**Comment by yanhuacuo** [comment]
目前我们自己基于 Qt 开发的软件已经完工：https://wubi98.gitee.io/2023/06/16/2023-06-16-01.rimetool/

librime 部分的功能，划给了「rimetool」，而 weasel 纯前端的功能，划给了「weasel-tool」。

暂时没有做其它平台的编译分发，「weasel-tool」的实现方法，可以简单改一改代码变成「squirrel-tool」。

但是这个工具，仅适用于五笔。 soak 已经实现的功能，重新梳理打包的话，肯定比我们的更通用，也更易用。

大佬加油。
